File: haskell/higher_order_functions/HigherOrderFunctions.hs
-- The (Int -> Int -> Int) is a higher order function.

-- It states that this must be a function, that takes two Int and return 
--   an Int, there op takes that function as an argument as well as two
--   other Ints as arguments to op.

op :: (Int -> Int -> Int) -> Int -> Int -> Int 
op f x y = f x y

prod2 :: Int -> Int -> Int
prod2 x y = x * y

main = do
          putStrLn(show(op (\x y -> x + y + 1) 4 5))
          putStrLn $ show $ op (\x y -> x + y + 1) 4 5
          putStrLn $ show $ op sum2 4 5

File: clojure/higher_order_functions/higher_order_functions.clj
;; A higher order function is a function that takes another function.
;; In this example we have a function op that takes a function f.
;; Notice that the f is taken from the vector argument list and used as an 
;; the first element of the list since it is a function that is be evaluated.

(defn op [f x y] (f x y))

;; A declared function
(defn prod2 [x y] (* x y))

;; Using an anonymous function, see anonymous_functions/anonymous_functions.clj
(assert (= (op (fn [x y] (+ x y 1)) 4 5) 10))

;; Shorthand for an anonymous function
(assert (= (op #(+ %1 %2 1) 4 5) 10))

;; Even shorter hand of anonymous function, where as the % is the first parameter
(assert (= (op #(+ % %2 1) 4 5) 10))

;; Using the prod2 function as the higher order function
(assert (= (op prod2 4 5) 20))
